### PR TITLE
Support Dashboard replicas (#2899)

### PR DESCRIPTION
**Pull request by KIVagant**
This PR makes possible to increase the amount of web dashboard replicas.

Follows up #2899

Signed-off-by: Eugene Glotov <kivagant@gmail.com>

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -433,6 +433,7 @@ golang.org/x/tools v0.0.0-20190506145303-2d16b83fe98c/go.mod h1:RgjU9mgBXZiqYHBn
 golang.org/x/tools v0.0.0-20190524140312-2c0ae7006135/go.mod h1:RgjU9mgBXZiqYHBnxXauZ1Gv1EHHAz9KjViQ78xBX0Q=
 golang.org/x/tools v0.0.0-20191009213438-b090f1f24028 h1:uYpiJNFQR6EGzWevFLoFS3mfOVJdQce29o6eRxtYqYA=
 golang.org/x/tools v0.0.0-20191009213438-b090f1f24028/go.mod h1:b+2E5dAYhXwXZwtnZ6UAqBI28+e2cm9otk0dWdXHAEo=
+golang.org/x/tools v0.0.0-20191024074452-7defa796fec0 h1:G/HlsVBRpvScWBPC2N5F3t6KORPLQD6n5k2nPsWt6DU=
```
Comment (KIVagant) [inline go.sum]: No idea where the change came from

**Comment by adleong** [comment]
Thanks for submitting this, @KIVagant!  I've got a few questions as I try to understand the issue:

Is sessionAffinity irrelevant when there is only one replica?  If so, should we just default the sessionAffinity to ClientIP all of the time?

In what types of scenarios would you need to change the sessionAffinity to something else?  Are these scenarios uncommon enough that we don't need to add a flag for configuring it? (users can always use something like kustomize if they really need to change it)

**Comment by KIVagant** [comment]
With only one replica sessionAffinity changes really nothing because there's no real balancing. But yes, overall it should be enabled by default. I decided not to set it by default to ClientIP to not affect any existing installations. If not only me believe the sessionAffinity setting is useful for the dashboard service, I can enable it by default.

**Comment by adleong** [comment]
Unless there's some downside to having sessionAffinity as ClientIP when there is only one replica, I would say let's make it the default.

**Comment by KIVagant** [comment]
@adleong ,  sessionAffinity and sessionAffinityConfig are enabled by default.

**Comment by grampelberg** [comment]
I don't think this is harmful, but I'm also pretty sure it won't actually fix anything. `port-forward` is always to a specific pod, so that would skip all this. For anything going through an ingress controller, they'll use endpoints and the stickiness would be required there.

**Comment by adleong** [comment]
This is a good point.  Which ingress are you testing this with, @KIVagant?

**Comment by KIVagant** [comment]
@grampelberg , it is very possible that I missed or confused something, there is a lot of settings involved. I'm using ALB Ingress Controller which talks to [oauth2_proxy](https://github.com/pusher/oauth2_proxy) which in its turn talks to Linkerd dashboard. The ALB Ingress Controller supports stickness options, for example `stickness.enabled=true,stinkyness.ib_cookie.duration_seconds=1200`. So it sticks to the oauth2_proxy pod(s) but I _feel_ like oauth2_proxy may have some problems if Linkerd dashboard has 2 or more pods but the service does not have the enabled sessionAffinity. I will try to make more tests.

**Comment by grampelberg** [comment]
@KIVagant I hadn't even thought about `oauth2_proxy`! Mind sharing your config for that? I bet this'll fix that exact use case.

**Comment by KIVagant** [comment]
The only setting they have is called [`-proxy-websockets`](https://pusher.github.io/oauth2_proxy/configuration) and I set it to `true`. Together with the stickness settings for ALB it works for one pod. But I'm not sure if it still the same for 2 and more pods (don't have the access to the cluster atm, will check tomorrow).

**Comment by KIVagant** [comment]
I did some testing. Prerequisites: stickness enabled for the ALB. The dashboard scaled to 3 replicas.

Here's the results:

Pod A:
image

Pod B:
image

After `sessionAffinity` is enabled, external requests from `po/oauth2-proxy-linkerd` disappeared from `linkerd tap pod/linkerd-web-A --namespace linkerd` but continued reaching the Pod B. 

So from one side, stickness is working only between the ALB to the oauth2-proxy. From the other side, I don't see any errors in Linkerd dashboard when the sessionAffinity is disabled. Taking this into account I would say this PR is not important unless someone report about a problem.

**Comment by KIVagant** [comment]
Btw I can extract the `Replicas` option to another PR.

**Comment by grampelberg** [comment]
I don't mind it going in, but I'd like to remove all configuration for `SessionAffinity`. It'll probably just be easier to have a very small PR for changing the replicas count.

**Comment by KIVagant** [comment]
Affinity removed

**Comment by KIVagant** [comment]
> No spaces to match overall style, and the default is not really required

@alpeb , done

File: go.sum (hunk 1 above)
Comment (kivagant-ba) [reply go.sum]: Removed

**Comment by KIVagant** [comment]
Thank you for merging the PR and helping me to analyze the use case. The result is tiny but for me it was a good example to practice with the repo and with Golang tests. Sorry for the noise produced by this.

**Comment by grampelberg** [comment]
Thank you for the contribution! It was super helpful to think things through on our side as well =)

**Pull request merged**